salt/cloud/clouds/profitbricks.py

```diff
@@ -0,0 +1,836 @@
+            'when trying to run the initial deployment: \n{1}'.format(
+                vm_['name'], exc
+            ),
+            exc_info_on_loglevel=logging.DEBUG
+        )
+        return False
+
+    create_network_interfaces(conn, datacenter_id, data['id'], vm_)
+    set_public_lan(conn, vm_)
+
+    def __query_node_data(vm_, data):
+        running = False
+        try:
+            data = show_instance(vm_['name'], 'action')
+            if not data:
+                return False
+            log.debug(
+                'Loaded node data for {0}:\nname: {1}\nstate: {2}'.format(
+                    vm_['name'],
+                    pprint.pformat(data['name']),
+                    data['vmState']
+                )
+            )
+        except Exception as err:
+            log.error(
+                'Failed to get nodes list: {0}'.format(
+                    err
+                ),
+                # Show the trackback if the debug logging level is enabled
+                exc_info_on_loglevel=logging.DEBUG
+            )
+            # Trigger a failure in the wait for IP function
+            return False
+
+        running = data['vmState'] == 'RUNNING'
+        if not running:
+            # Still not running, trigger another iteration
+            return
+
+        if ssh_interface(vm_) == 'private_lan' and data['private_ips']:
+            vm_['ssh_host'] = data['private_ips'][0]
+
+        if ssh_interface(vm_) != 'private_lan' and data['public_ips']:
+            vm_['ssh_host'] = data['public_ips'][0]
+
+        return data
+
+    try:
+        data = salt.utils.cloud.wait_for_ip(
+            __query_node_data,
+            update_args=(vm_, data),
+            timeout=config.get_cloud_config_value(
+                'wait_for_ip_timeout', vm_, __opts__, default=10 * 60),
+            interval=config.get_cloud_config_value(
+                'wait_for_ip_interval', vm_, __opts__, default=10),
+        )
+    except (SaltCloudExecutionTimeout, SaltCloudExecutionFailure) as exc:
+        try:
+            # It might be already up, let's destroy it!
+            destroy(vm_['name'])
+        except SaltCloudSystemExit:
+            pass
+        finally:
+            raise SaltCloudSystemExit(str(exc.message))
+
+    log.debug('VM is now running')
+    log.info('Created Cloud VM {0[name]!r}'.format(vm_))
+    log.debug(
+        '{0[name]!r} VM creation details:\n{1}'.format(
+            vm_, pprint.pformat(data)
+        )
+    )
+
+    salt.utils.cloud.fire_event(
+        'event',
+        'created instance',
+        'salt/cloud/{0}/created'.format(vm_['name']),
+        {
+            'name': vm_['name'],
+            'profile': vm_['profile'],
+            'provider': vm_['driver'],
+        },
+        transport=__opts__['transport']
+    )
+
+    if 'ssh_host' in vm_:
+        vm_['key_filename'] = get_key_filename(vm_)
+        ret = salt.utils.cloud.bootstrap(vm_, __opts__)
```

Comment:
I think it would be better to keep things consistent with the other drivers. I don't think we should say "success!" if a user is expecting the VM to be bootstrapped. It also looks like passwords are supported here too? If someone provides a password instead of an ssh_key, that seems like it should work for bootstrapping. We actually handle this error messaging in the bootstrap function itself, right at the beginning: https://github.com/saltstack/salt/blob/develop/salt/utils/cloud.py#L292.

One thing to note is that there is a `no-deploy` option that can be passed to the salt-cloud command either from the CLI with `--no-deploy`, or from one of the salt-cloud config files with `deploy: False`. Essentially what this does is it acts like you have this built right now - it will create an instance, but it won't deploy/run the bootstrap script to install salt. Perhaps that would be a good way to address this scenario. If the VM can't be bootstrapped because a keyfile, ssh_key, or password is missing, or because you want to maintain the "provision but don't install salt" behavior, you could add a note to the docs to use `deploy: False`. That way we can keep behavior consistent across providers, and not be confusing to users.